ledger/sandbox/src/main/scala/com/digitalasset/platform/store/dao/events/TransactionsWriter.scala

```diff
@@ -41,16 +45,62 @@ private[dao] object TransactionsWriter extends TransactionsWriter {
     }
   }
 
-  def apply(
+  private def divulgedContracts(
+      disclosure: DisclosureRelation,
+      toBeInserted: Set[ContractId],
+  ): PartialFunction[(NodeId, Node), (ContractId, Set[Party])] = {
+    case (nodeId, c: Create) if toBeInserted(c.coid) =>
+      c.coid -> disclosure(nodeId)
+    case (nodeId, e: Exercise) if toBeInserted(e.targetCoid) =>
+      e.targetCoid -> disclosure(nodeId)
+    case (nodeId, f: Fetch) if toBeInserted(f.coid) =>
+      f.coid -> disclosure(nodeId)
+    case (nodeId, l: LookupByKey) if l.result.fold(false)(toBeInserted) =>
+      l.result.get -> disclosure(nodeId)
+  }
+
+  private def divulgence(
+      transaction: Transaction,
+      disclosure: DisclosureRelation,
+      toBeInserted: Set[ContractId],
+  ): WitnessRelation[ContractId] =
+    if (toBeInserted.isEmpty) {
+      Map.empty
+    } else {
+      transaction.nodes.iterator
+        .collect(divulgedContracts(disclosure, toBeInserted))
+        .foldLeft[WitnessRelation[ContractId]](Map.empty)(Relation.merge)
+    }
+
+  private def prepareWitnessesBatch(
+      insertions: Set[ContractId],
+      deletions: Set[ContractId],
+      transaction: Transaction,
+      blinding: BlindingInfo,
+  ): Option[BatchSql] = {
+    val localDivulgence = divulgence(transaction, blinding.disclosure, insertions)
+    val fullDivulgence = Relation.union(
```

Comment:
I'm not sure I understand this correctly. Do you observe a difference between this `localDivulgence` and `blindingInfo.globalDivulgence`?

Reply:
Yes, `blindingInfo.globalDivulgence` does not include the divulgence relation between contracts that are being divulged in the current transaction. If I only use `globalDivulgence` I will not add the witnesses of contracts divulged locally.